src/main/java/org/broadinstitute/hellbender/utils/downsampling/ReadsDownsamplingIterator.java

```diff
@@ -0,0 +1,91 @@
+package org.broadinstitute.hellbender.utils.downsampling;
+
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * Iterator wrapper around our generic {@link ReadsDownsampler)} interface. Converts the push-style
+ * downsampler interface to a pull model.
+ */
+public final class ReadsDownsamplingIterator implements Iterator<GATKRead>, Iterable<GATKRead> {
+
+    private final Iterator<GATKRead> nestedReadIterator;
+    private final ReadsDownsampler downsampler;
+    private Collection<GATKRead> downsampledReadsCache;
+    private GATKRead nextRead = null;
+    private Iterator<GATKRead> downsampledReadsCacheIterator = null;
+
+    /**
+     * @param iter wrapped iterator from which this iterator will pull reads
```

Comment:
Not sure what you mean here -- this parameter is simply an iterator of reads that are to be downsampled, as the docs state. What more do you want me to say?